File: algorithms/algorithms-java/src/main/java/com/lanceolata/coding/algorithms/java/leetcode/Question_0401_Binary_Watch.java
package com.lanceolata.coding.algorithms.java.leetcode;

import java.util.ArrayList;
import java.util.List;

public class Question_0401_Binary_Watch {
    public List<String> readBinaryWatch(int turnedOn) {
        List<String> res = new ArrayList<>();
        if (turnedOn < 0 || turnedOn > 12) {
            return res;
        }
        for (int h = 0; h < 12; h++) {
            for (int m = 0; m < 60; m++) {
                if (Integer.bitCount((h << 6) + m) == turnedOn) {
                    res.add(String.format("%d:%02d", h, m));
                }
            }
        }
        return res;
    }
}

File: algorithms/algorithms-cplusplus/leetcode/Question_0401_Binary_Watch.cc
#include <string>
#include <vector>

using namespace std;

class Solution {
 public:
  vector<string> readBinaryWatch(int turnedOn) {
    vector<string> res;
    if (turnedOn < 0 || turnedOn > 10) {
      return res;
    }
    for (int h = 0; h < 12; h++) {
      for (int m = 0; m < 60; m++) {
        int n = (h << 6) | m;
        if (bitCount(n) == turnedOn) {
          res.emplace_back(to_string(h) + (m < 10 ? ":0" : ":") + to_string(m));
        }
      }
    }
    return res;
  }

  int bitCount(int n) {
    int res = 0;
    while (n != 0) {
      res++;
      n &= n - 1;
    }
    return res;
  }
};
